### PR TITLE
TestCaseFilter with custom properties

### PR DESCRIPTION
**Pull request by engyebrahim**
fix: https://github.com/microsoft/testfx/issues/731
fix: https://github.com/microsoft/testfx/issues/201

### CODE REVIEW

File: src/Adapter/MSTest.TestAdapter/TestMethodFilter.cs
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/TestMethodFilter.cs:97]: I think this one will only work when running tests but not discovering tests. 

Also, I don't know if we should be returning the value somehow for Test Explorer UI (cc @nohwnd @drognanar )

Comment (engyebrahim) [reply src/Adapter/MSTest.TestAdapter/TestMethodFilter.cs:97]: image
I tested it and discovering(--list-tests) is working correctly

Comment (engyebrahim) [reply src/Adapter/MSTest.TestAdapter/TestMethodFilter.cs:97]: and for test explorer all appears here using mstest dev-pack with this change
image

Comment (Evangelink) [reply src/Adapter/MSTest.TestAdapter/TestMethodFilter.cs:97]: Ok cool then I have absolutely no idea how this works. If you look at `GetFilterExpression` it calls `GetTestCaseFilterFromDiscoveryContext` that calls `PropertyProvider` but never the method you have edited.

I think you can go on as it seems to be working but I would love to have a review from @nohwnd before this gets merged.

File: test/IntegrationTests/MSTest.Acceptance.IntegrationTests/TestFilterTests.cs
Comment (nohwnd) [inline test/IntegrationTests/MSTest.Acceptance.IntegrationTests/TestFilterTests.cs:87]: What are these tests proving? The errors seem unrelated to adding a new functionality that can filter on additional test categories.

File: src/Adapter/MSTest.TestAdapter/TestMethodFilter.cs
Comment (nohwnd) [reply src/Adapter/MSTest.TestAdapter/TestMethodFilter.cs:97]:
```
>	Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.dll!Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.TestMethodFilter.PropertyValueProvider(Microsoft.VisualStudio.TestPlatform.ObjectModel.TestCase currentTest = {Microsoft.VisualStudio.TestPlatform.ObjectModel.TestCase}, string propertyName = "tree") Line 97	C#
 	Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.dll!Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.UnitTestDiscoverer.SendTestCases.AnonymousMethod__0(string p = "tree") Line 114	C#
 	Microsoft.Testing.Extensions.VSTestBridge.dll!Microsoft.Testing.Extensions.VSTestBridge.ObjectModel.FastFilter.TryGetPropertyValue(string name = "tree", System.Func<string, object> propertyValueProvider = {Method = {System.Reflection.RuntimeMethodInfo}}, out string singleValue = null, out string[] multiValues = null) Line 214	C#
 	Microsoft.Testing.Extensions.VSTestBridge.dll!Microsoft.Testing.Extensions.VSTestBridge.ObjectModel.FastFilter.Evaluate(System.Func<string, object> propertyValueProvider = {Method = {System.Reflection.RuntimeMethodInfo}}) Line 159	C#
 	Microsoft.Testing.Extensions.VSTestBridge.dll!Microsoft.Testing.Extensions.VSTestBridge.ObjectModel.FilterExpressionWrapper.Evaluate(System.Func<string, object> propertyValueProvider = {Method = {System.Reflection.RuntimeMethodInfo}}) Line 89	C#
 	Microsoft.Testing.Extensions.VSTestBridge.dll!Microsoft.Testing.Extensions.VSTestBridge.ObjectModel.TestCaseFilterExpression.MatchTestCase(Microsoft.VisualStudio.TestPlatform.ObjectModel.TestCase testCase = {Microsoft.VisualStudio.TestPlatform.ObjectModel.TestCase}, System.Func<string, object> propertyValueProvider = {Method = {System.Reflection.RuntimeMethodInfo}}) Line 47	C#
 	Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.dll!Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.UnitTestDiscoverer.SendTestCases(string source = "S:\\t\\mstest13_\\prj\\bin\\Debug\\net8.0\\mstest133.dll", System.Collections.Generic.IEnumerable<Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.ObjectModel.UnitTestElement> testElements = Count = 2, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.ITestCaseDiscoverySink discoverySink = {Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.TestCaseDiscoverySink}, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IDiscoveryContext discoveryContext = {Microsoft.Testing.Extensions.VSTestBridge.ObjectModel.RunContextAdapter}, Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging.IMessageLogger logger = {Microsoft.Testing.Extensions.VSTestBridge.ObjectModel.FrameworkHandlerAdapter}) Line 114	C#
 	Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.dll!Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.UnitTestDiscoverer.DiscoverTestsInSource(string source = "S:\\t\\mstest13_\\prj\\bin\\Debug\\net8.0\\mstest133.dll", Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging.IMessageLogger logger = {Microsoft.Testing.Extensions.VSTestBridge.ObjectModel.FrameworkHandlerAdapter}, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.ITestCaseDiscoverySink discoverySink = {Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.TestCaseDiscoverySink}, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IDiscoveryContext discoveryContext = {Microsoft.Testing.Extensions.VSTestBridge.ObjectModel.RunContextAdapter}) Line 87	C#
 	Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.dll!Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.Execution.TestExecutionManager.RunTests(System.Collections.Generic.IEnumerable<string> sources = {string[1]}, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IRunContext runContext = {Microsoft.Testing.Extensions.VSTestBridge.ObjectModel.RunContextAdapter}, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IFrameworkHandle frameworkHandle = {Microsoft.Testing.Extensions.VSTestBridge.ObjectModel.FrameworkHandlerAdapter}, Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.TestRunCancellationToken cancellationToken = {Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.TestRunCancellationToken}) Line 112	C#
 	Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.dll!Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.MSTestExecutor.RunTests.AnonymousMethod__0(Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.TestRunCancellationToken testRunToken = {Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.TestRunCancellationToken}) Line 72	C#
 	Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.dll!Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.MSTestExecutor.RunTestsFromRightContext.__DoRunTests|0() Line 126	C#
 	Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.dll!Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.MSTestExecutor.RunTestsFromRightContext(Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IFrameworkHandle frameworkHandle = {Microsoft.Testing.Extensions.VSTestBridge.ObjectModel.FrameworkHandlerAdapter}, System.Action<Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.TestRunCancellationToken> runTestsAction = {Method = {System.Reflection.RuntimeMethodInfo}}) Line 115	C#
 	Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.dll!Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.MSTestExecutor.RunTests(System.Collections.Generic.IEnumerable<string> sources = {string[1]}, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IRunContext runContext = {Microsoft.Testing.Extensions.VSTestBridge.ObjectModel.RunContextAdapter}, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IFrameworkHandle frameworkHandle = {Microsoft.Testing.Extensions.VSTestBridge.ObjectModel.FrameworkHandlerAdapter}) Line 72	C#
 	Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter.dll!Microsoft.VisualStudio.TestTools.UnitTesting.MSTestBridgedTestFramework.SynchronizedRunTestsAsync(Microsoft.Testing.Extensions.VSTestBridge.Requests.VSTestRunTestExecutionRequest request = {Microsoft.Testing.Extensions.VSTestBridge.Requests.VSTestRunTestExecutionRequest}, Microsoft.Testing.Platform.Messages.IMessageBus messageBus = {Microsoft.Testing.Platform.Messages.MessageBusProxy}, System.Threading.CancellationToken cancellationToken = IsCancellationRequested = false) Line 56	C#
 	Microsoft.Testing.Extensions.VSTestBridge.dll!Microsoft.Testing.Extensions.VSTestBridge.SynchronizedSingleSessionVSTestBridgedTestFramework.ExecuteRequestAsync.AnonymousMethod__0() Line 129	C#
 	Microsoft.Testing.Extensions.VSTestBridge.dll!Microsoft.Testing.Extensions.VSTestBridge.SynchronizedSingleSessionVSTestBridgedTestFramework.ExecuteRequestWithRequestCountGuardAsync(System.Func<System.Threading.Tasks.Task> asyncFunc = {Method = {System.Reflection.RuntimeMethodInfo}}) Line 150	C#
 	Microsoft.Testing.Extensions.VSTestBridge.dll!Microsoft.Testing.Extensions.VSTestBridge.SynchronizedSingleSessionVSTestBridgedTestFramework.ExecuteRequestAsync(Microsoft.Testing.Platform.Requests.TestExecutionRequest request = {Microsoft.Testing.Platform.Requests.RunTestExecutionRequest}, Microsoft.Testing.Platform.Messages.IMessageBus messageBus = {Microsoft.Testing.Platform.Messages.MessageBusProxy}, System.Threading.CancellationToken cancellationToken = IsCancellationRequested = false) Line 118	C#
 	Microsoft.Testing.Extensions.VSTestBridge.dll!Microsoft.Testing.Extensions.VSTestBridge.VSTestBridgedTestFrameworkBase.ExecuteRequestAsync(Microsoft.Testing.Platform.Extensions.TestFramework.ExecuteRequestContext context = {Microsoft.Testing.Platform.Extensions.TestFramework.ExecuteRequestContext}) Line 58	C#
 	Microsoft.Testing.Platform.dll!Microsoft.Testing.Platform.Requests.TestHostTestFrameworkInvoker.ExecuteRequestAsync(Microsoft.Testing.Platform.Extensions.TestFramework.ITestFramework testFramework = {Microsoft.VisualStudio.TestTools.UnitTesting.MSTestBridgedTestFramework}, Microsoft.Testing.Platform.Requests.TestExecutionRequest request = {Microsoft.Testing.Platform.Requests.RunTestExecutionRequest}, Microsoft.Testing.Platform.Messages.IMessageBus messageBus = {Microsoft.Testing.Platform.Messages.MessageBusProxy}, System.Threading.CancellationToken cancellationToken = IsCancellationRequested = false) Line 73	C#
 	Microsoft.Testing.Platform.dll!Microsoft.Testing.Platform.Requests.TestHostTestFrameworkInvoker.ExecuteAsync(Microsoft.Testing.Platform.Extensions.TestFramework.ITestFramework testFramework = {Microsoft.VisualStudio.TestTools.UnitTesting.MSTestBridgedTestFramework}, Microsoft.Testing.Platform.TestHost.ClientInfo client = {Microsoft.Testing.Platform.TestHost.ClientInfo}, System.Threading.CancellationToken cancellationToken = IsCancellationRequested = false) Line 62	C#
 	Microsoft.Testing.Platform.dll!Microsoft.Testing.Platform.Hosts.CommonTestHost.ExecuteRequestAsync(Microsoft.Testing.Platform.OutputDevice.IPlatformOutputDevice outputDevice = {Microsoft.Testing.Platform.OutputDevice.ConsoleOutputDevice}, Microsoft.Testing.Platform.Services.ITestSessionContext testSessionInfo = {Microsoft.Testing.Platform.Services.TestSessionContext}, Microsoft.Testing.Platform.Services.ServiceProvider serviceProvider = {Microsoft.Testing.Platform.Services.ServiceProvider}, Microsoft.Testing.Platform.Messages.BaseMessageBus baseMessageBus = {Microsoft.Testing.Platform.Messages.MessageBusProxy}, Microsoft.Testing.Platform.Extensions.TestFramework.ITestFramework testFramework = {Microsoft.VisualStudio.TestTools.UnitTesting.MSTestBridgedTestFramework}, Microsoft.Testing.Platform.TestHost.ClientInfo client = {Microsoft.Testing.Platform.TestHost.ClientInfo}) Line 75	C#
 	Microsoft.Testing.Platform.dll!Microsoft.Testing.Platform.Hosts.ConsoleTestHost.InternalRunAsync() Line 101	C#
 	Microsoft.Testing.Platform.dll!Microsoft.Testing.Platform.Hosts.CommonTestHost.RunAsync() Line 38	C#
 	Microsoft.Testing.Platform.dll!Microsoft.Testing.Platform.Builder.TestApplication.RunAsync() Line 251	C#
 	mstest133.dll!TestingPlatformEntryPoint.Main(string[] args = {string[2]}) Line 20	C#
 	mstest133.dll!TestingPlatformEntryPoint.<Main>(string[] args = {string[2]})	Unknown

```

This is the stack.

Comment (nohwnd) [reply src/Adapter/MSTest.TestAdapter/TestMethodFilter.cs:97]: My concern here is speed, Traits are internally properties on the TestObject, and retrieving them is not trivial. It would be interesting to see the impact of this on 10k tests, where none, 1/2 and all have some traits.

https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.ObjectModel/TraitCollection.cs#L68

File: test/IntegrationTests/MSTest.Acceptance.IntegrationTests/TestFilterTests.cs
Comment (engyebrahim) [reply test/IntegrationTests/MSTest.Acceptance.IntegrationTests/TestFilterTests.cs:87]: yup it's not related to the fix, I just want to document that behaviour as it's related to `TestProperty` if would change it in future

File: src/Adapter/MSTest.TestAdapter/TestMethodFilter.cs
Comment (engyebrahim) [reply src/Adapter/MSTest.TestAdapter/TestMethodFilter.cs:97]: Saving the state as me and @nohwnd discussed offline.
I tested it with 10k test having `[TestProperty]` and with filter matching  and the performance were to close to the old behaviour

Comment (Evangelink) [reply src/Adapter/MSTest.TestAdapter/TestMethodFilter.cs:97]: Awesome! Thanks @engyebrahim

**Review by nohwnd — APPROVED** [review]
I think this is useful functionality and it aligns well with traits on tests being used as *properties* of the test.

There is some slight perf impact when the property filter is enabled, but it is not the mainline scenario (discovery under VS), and we should be able to implement this in a more optimized way with ease if we drop object model from mstest.

**Comment by Evangelink** [comment]
Nice work @engyebrahim :shipit:

**Pull request merged**
